doc/developer-guide/internal-libraries/Extendible.en.rst

```diff
@@ -0,0 +1,129 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+   or more contributor license agreements.  See the NOTICE file
+   distributed with this work for additional information
+   regarding copyright ownership.  The ASF licenses this file
+   to you under the Apache License, Version 2.0 (the
+   "License"); you may not use this file except in compliance
+   with the License.  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing,
+   software distributed under the License is distributed on an
+   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+   KIND, either express or implied.  See the License for the
+   specific language governing permissions and limitations
+   under the License.
+
+.. include:: ../../common.defs
+
+.. highlight:: cpp
+.. default-domain:: cpp
+
+Extendible
+**********
+
+Synopsis
+++++++++
+
+.. code-block:: cpp
+
+   #include "ts/Extendible.h"
+
+Allow code (and Plugins) to declare member variables of a class during system init. Like metadata, but allocated and accessed more efficiently, and staticly.
+
+Provide API for atomic and thread safe field access, to reduce the complexity and lock contention of it's container and user code.
+
+Description
++++++++++++
+
+A data class that inherits from Extendible, uses a CRTP (Curiously Recursive Template Pattern) so that it's static Schema instance is unique from other Extendible types.
+
+.. code-block:: cpp
+
+   class ExtendibleExample : Extendible<ExtendibleExample> {
+      int real_member = 0;
+   }
+
+Structures
+----------
+* FieldId<AccessType_, FieldType> - the handle used to access a field. These are templated to prevent human error, and branching logic.
+* FieldId_C - the handle used to access a field through C API. Human error not allowed by convention.
+* FieldSchema - stores attributes, constructor and destructor of a field.
+* Schema - manages fields and memory layout of an Extendible type. 
+* Extendible<DerivedType> - allocates block of memory, uses FieldId or schema to access slices of memory.
+
+
+During system init, code and plugins can add fields to the Extendible's schema. This will update the `Memory Layout`_ of the schema, and the memory offsets of all fields. The schema does not know the FieldType, but it stores the byte size and creates lambdas of the type's constructor and destructor. And to avoid corruption, the code asserts that no instances are in use when adding fields. 
+
+.. code-block:: cpp
+
+   ExtendibleExample:FieldId<ATOMIC,int> fld_my_int;
+   
+   void PluginInit() {
+     fld_my_int = ExtendibleExample::schema.addField("my_plugin_int");
+   }
+
+
+When an Extendible derived class is instantiated, new() will allocate a block of memory for the derived class and all added fields. There is zero memory overhead per instance, unless using :ref:ACIDPTR field access type.
```

Comment:
* `new()` needs markup
* markup for `ACIDPTR` seems broken
* `new() will allocate a block of memory`...`There is zero memory overhead per instance` What!?

Reply:
I think Aaron means "zero overhead for meta-data". That is, all of the schema information is stored elsewhere, either in global data or the compiler symbol table. But not per instance.